// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: rill/local/v1/api.proto

package localv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on PingRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PingRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PingRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PingRequestMultiError, or
// nil if none found.
func (m *PingRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *PingRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return PingRequestMultiError(errors)
	}

	return nil
}

// PingRequestMultiError is an error wrapping multiple validation errors
// returned by PingRequest.ValidateAll() if the designated constraints aren't met.
type PingRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PingRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PingRequestMultiError) AllErrors() []error { return m }

// PingRequestValidationError is the validation error returned by
// PingRequest.Validate if the designated constraints aren't met.
type PingRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PingRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PingRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PingRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PingRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PingRequestValidationError) ErrorName() string { return "PingRequestValidationError" }

// Error satisfies the builtin error interface
func (e PingRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPingRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PingRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PingRequestValidationError{}

// Validate checks the field values on PingResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *PingResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PingResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in PingResponseMultiError, or
// nil if none found.
func (m *PingResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *PingResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTime()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, PingResponseValidationError{
					field:  "Time",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, PingResponseValidationError{
					field:  "Time",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTime()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return PingResponseValidationError{
				field:  "Time",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return PingResponseMultiError(errors)
	}

	return nil
}

// PingResponseMultiError is an error wrapping multiple validation errors
// returned by PingResponse.ValidateAll() if the designated constraints aren't met.
type PingResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PingResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PingResponseMultiError) AllErrors() []error { return m }

// PingResponseValidationError is the validation error returned by
// PingResponse.Validate if the designated constraints aren't met.
type PingResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PingResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PingResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PingResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PingResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PingResponseValidationError) ErrorName() string { return "PingResponseValidationError" }

// Error satisfies the builtin error interface
func (e PingResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPingResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PingResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PingResponseValidationError{}

// Validate checks the field values on GetMetadataRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetMetadataRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetMetadataRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetMetadataRequestMultiError, or nil if none found.
func (m *GetMetadataRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetMetadataRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GetMetadataRequestMultiError(errors)
	}

	return nil
}

// GetMetadataRequestMultiError is an error wrapping multiple validation errors
// returned by GetMetadataRequest.ValidateAll() if the designated constraints
// aren't met.
type GetMetadataRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetMetadataRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetMetadataRequestMultiError) AllErrors() []error { return m }

// GetMetadataRequestValidationError is the validation error returned by
// GetMetadataRequest.Validate if the designated constraints aren't met.
type GetMetadataRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetMetadataRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetMetadataRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetMetadataRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetMetadataRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetMetadataRequestValidationError) ErrorName() string {
	return "GetMetadataRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetMetadataRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetMetadataRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetMetadataRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetMetadataRequestValidationError{}

// Validate checks the field values on GetMetadataResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetMetadataResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetMetadataResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetMetadataResponseMultiError, or nil if none found.
func (m *GetMetadataResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetMetadataResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for InstanceId

	// no validation rules for ProjectPath

	// no validation rules for InstallId

	// no validation rules for UserId

	// no validation rules for Version

	// no validation rules for BuildCommit

	// no validation rules for BuildTime

	// no validation rules for IsDev

	// no validation rules for AnalyticsEnabled

	// no validation rules for Readonly

	// no validation rules for GrpcPort

	// no validation rules for LoginUrl

	// no validation rules for AdminUrl

	if len(errors) > 0 {
		return GetMetadataResponseMultiError(errors)
	}

	return nil
}

// GetMetadataResponseMultiError is an error wrapping multiple validation
// errors returned by GetMetadataResponse.ValidateAll() if the designated
// constraints aren't met.
type GetMetadataResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetMetadataResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetMetadataResponseMultiError) AllErrors() []error { return m }

// GetMetadataResponseValidationError is the validation error returned by
// GetMetadataResponse.Validate if the designated constraints aren't met.
type GetMetadataResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetMetadataResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetMetadataResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetMetadataResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetMetadataResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetMetadataResponseValidationError) ErrorName() string {
	return "GetMetadataResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetMetadataResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetMetadataResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetMetadataResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetMetadataResponseValidationError{}

// Validate checks the field values on GetVersionRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetVersionRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetVersionRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetVersionRequestMultiError, or nil if none found.
func (m *GetVersionRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetVersionRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GetVersionRequestMultiError(errors)
	}

	return nil
}

// GetVersionRequestMultiError is an error wrapping multiple validation errors
// returned by GetVersionRequest.ValidateAll() if the designated constraints
// aren't met.
type GetVersionRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetVersionRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetVersionRequestMultiError) AllErrors() []error { return m }

// GetVersionRequestValidationError is the validation error returned by
// GetVersionRequest.Validate if the designated constraints aren't met.
type GetVersionRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetVersionRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetVersionRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetVersionRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetVersionRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetVersionRequestValidationError) ErrorName() string {
	return "GetVersionRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetVersionRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetVersionRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetVersionRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetVersionRequestValidationError{}

// Validate checks the field values on GetVersionResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetVersionResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetVersionResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetVersionResponseMultiError, or nil if none found.
func (m *GetVersionResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetVersionResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Current

	// no validation rules for Latest

	if len(errors) > 0 {
		return GetVersionResponseMultiError(errors)
	}

	return nil
}

// GetVersionResponseMultiError is an error wrapping multiple validation errors
// returned by GetVersionResponse.ValidateAll() if the designated constraints
// aren't met.
type GetVersionResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetVersionResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetVersionResponseMultiError) AllErrors() []error { return m }

// GetVersionResponseValidationError is the validation error returned by
// GetVersionResponse.Validate if the designated constraints aren't met.
type GetVersionResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetVersionResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetVersionResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetVersionResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetVersionResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetVersionResponseValidationError) ErrorName() string {
	return "GetVersionResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetVersionResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetVersionResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetVersionResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetVersionResponseValidationError{}

// Validate checks the field values on GitStatusRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GitStatusRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GitStatusRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GitStatusRequestMultiError, or nil if none found.
func (m *GitStatusRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GitStatusRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GitStatusRequestMultiError(errors)
	}

	return nil
}

// GitStatusRequestMultiError is an error wrapping multiple validation errors
// returned by GitStatusRequest.ValidateAll() if the designated constraints
// aren't met.
type GitStatusRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GitStatusRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GitStatusRequestMultiError) AllErrors() []error { return m }

// GitStatusRequestValidationError is the validation error returned by
// GitStatusRequest.Validate if the designated constraints aren't met.
type GitStatusRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GitStatusRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GitStatusRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GitStatusRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GitStatusRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GitStatusRequestValidationError) ErrorName() string { return "GitStatusRequestValidationError" }

// Error satisfies the builtin error interface
func (e GitStatusRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGitStatusRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GitStatusRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GitStatusRequestValidationError{}

// Validate checks the field values on GitStatusResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GitStatusResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GitStatusResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GitStatusResponseMultiError, or nil if none found.
func (m *GitStatusResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GitStatusResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Branch

	// no validation rules for GithubUrl

	// no validation rules for ManagedGit

	// no validation rules for LocalChanges

	// no validation rules for LocalCommits

	// no validation rules for RemoteCommits

	if len(errors) > 0 {
		return GitStatusResponseMultiError(errors)
	}

	return nil
}

// GitStatusResponseMultiError is an error wrapping multiple validation errors
// returned by GitStatusResponse.ValidateAll() if the designated constraints
// aren't met.
type GitStatusResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GitStatusResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GitStatusResponseMultiError) AllErrors() []error { return m }

// GitStatusResponseValidationError is the validation error returned by
// GitStatusResponse.Validate if the designated constraints aren't met.
type GitStatusResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GitStatusResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GitStatusResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GitStatusResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GitStatusResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GitStatusResponseValidationError) ErrorName() string {
	return "GitStatusResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GitStatusResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGitStatusResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GitStatusResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GitStatusResponseValidationError{}

// Validate checks the field values on GitPullRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GitPullRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GitPullRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GitPullRequestMultiError,
// or nil if none found.
func (m *GitPullRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GitPullRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for DiscardLocal

	if len(errors) > 0 {
		return GitPullRequestMultiError(errors)
	}

	return nil
}

// GitPullRequestMultiError is an error wrapping multiple validation errors
// returned by GitPullRequest.ValidateAll() if the designated constraints
// aren't met.
type GitPullRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GitPullRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GitPullRequestMultiError) AllErrors() []error { return m }

// GitPullRequestValidationError is the validation error returned by
// GitPullRequest.Validate if the designated constraints aren't met.
type GitPullRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GitPullRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GitPullRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GitPullRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GitPullRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GitPullRequestValidationError) ErrorName() string { return "GitPullRequestValidationError" }

// Error satisfies the builtin error interface
func (e GitPullRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGitPullRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GitPullRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GitPullRequestValidationError{}

// Validate checks the field values on GitPullResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GitPullResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GitPullResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GitPullResponseMultiError, or nil if none found.
func (m *GitPullResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GitPullResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Output

	if len(errors) > 0 {
		return GitPullResponseMultiError(errors)
	}

	return nil
}

// GitPullResponseMultiError is an error wrapping multiple validation errors
// returned by GitPullResponse.ValidateAll() if the designated constraints
// aren't met.
type GitPullResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GitPullResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GitPullResponseMultiError) AllErrors() []error { return m }

// GitPullResponseValidationError is the validation error returned by
// GitPullResponse.Validate if the designated constraints aren't met.
type GitPullResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GitPullResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GitPullResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GitPullResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GitPullResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GitPullResponseValidationError) ErrorName() string { return "GitPullResponseValidationError" }

// Error satisfies the builtin error interface
func (e GitPullResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGitPullResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GitPullResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GitPullResponseValidationError{}

// Validate checks the field values on GitPushRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GitPushRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GitPushRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GitPushRequestMultiError,
// or nil if none found.
func (m *GitPushRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GitPushRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for CommitMessage

	// no validation rules for Force

	if len(errors) > 0 {
		return GitPushRequestMultiError(errors)
	}

	return nil
}

// GitPushRequestMultiError is an error wrapping multiple validation errors
// returned by GitPushRequest.ValidateAll() if the designated constraints
// aren't met.
type GitPushRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GitPushRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GitPushRequestMultiError) AllErrors() []error { return m }

// GitPushRequestValidationError is the validation error returned by
// GitPushRequest.Validate if the designated constraints aren't met.
type GitPushRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GitPushRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GitPushRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GitPushRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GitPushRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GitPushRequestValidationError) ErrorName() string { return "GitPushRequestValidationError" }

// Error satisfies the builtin error interface
func (e GitPushRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGitPushRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GitPushRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GitPushRequestValidationError{}

// Validate checks the field values on GitPushResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GitPushResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GitPushResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GitPushResponseMultiError, or nil if none found.
func (m *GitPushResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GitPushResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GitPushResponseMultiError(errors)
	}

	return nil
}

// GitPushResponseMultiError is an error wrapping multiple validation errors
// returned by GitPushResponse.ValidateAll() if the designated constraints
// aren't met.
type GitPushResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GitPushResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GitPushResponseMultiError) AllErrors() []error { return m }

// GitPushResponseValidationError is the validation error returned by
// GitPushResponse.Validate if the designated constraints aren't met.
type GitPushResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GitPushResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GitPushResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GitPushResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GitPushResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GitPushResponseValidationError) ErrorName() string { return "GitPushResponseValidationError" }

// Error satisfies the builtin error interface
func (e GitPushResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGitPushResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GitPushResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GitPushResponseValidationError{}

// Validate checks the field values on PushToGithubRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *PushToGithubRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PushToGithubRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// PushToGithubRequestMultiError, or nil if none found.
func (m *PushToGithubRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *PushToGithubRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Account

	// no validation rules for Repo

	if len(errors) > 0 {
		return PushToGithubRequestMultiError(errors)
	}

	return nil
}

// PushToGithubRequestMultiError is an error wrapping multiple validation
// errors returned by PushToGithubRequest.ValidateAll() if the designated
// constraints aren't met.
type PushToGithubRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PushToGithubRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PushToGithubRequestMultiError) AllErrors() []error { return m }

// PushToGithubRequestValidationError is the validation error returned by
// PushToGithubRequest.Validate if the designated constraints aren't met.
type PushToGithubRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PushToGithubRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PushToGithubRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PushToGithubRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PushToGithubRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PushToGithubRequestValidationError) ErrorName() string {
	return "PushToGithubRequestValidationError"
}

// Error satisfies the builtin error interface
func (e PushToGithubRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPushToGithubRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PushToGithubRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PushToGithubRequestValidationError{}

// Validate checks the field values on PushToGithubResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *PushToGithubResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on PushToGithubResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// PushToGithubResponseMultiError, or nil if none found.
func (m *PushToGithubResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *PushToGithubResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Remote

	// no validation rules for Account

	// no validation rules for Repo

	if len(errors) > 0 {
		return PushToGithubResponseMultiError(errors)
	}

	return nil
}

// PushToGithubResponseMultiError is an error wrapping multiple validation
// errors returned by PushToGithubResponse.ValidateAll() if the designated
// constraints aren't met.
type PushToGithubResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m PushToGithubResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m PushToGithubResponseMultiError) AllErrors() []error { return m }

// PushToGithubResponseValidationError is the validation error returned by
// PushToGithubResponse.Validate if the designated constraints aren't met.
type PushToGithubResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e PushToGithubResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e PushToGithubResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e PushToGithubResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e PushToGithubResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e PushToGithubResponseValidationError) ErrorName() string {
	return "PushToGithubResponseValidationError"
}

// Error satisfies the builtin error interface
func (e PushToGithubResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sPushToGithubResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = PushToGithubResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = PushToGithubResponseValidationError{}

// Validate checks the field values on DeployProjectRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeployProjectRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeployProjectRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeployProjectRequestMultiError, or nil if none found.
func (m *DeployProjectRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DeployProjectRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Org

	// no validation rules for NewOrgDisplayName

	// no validation rules for ProjectName

	// no validation rules for Upload

	// no validation rules for Archive

	if len(errors) > 0 {
		return DeployProjectRequestMultiError(errors)
	}

	return nil
}

// DeployProjectRequestMultiError is an error wrapping multiple validation
// errors returned by DeployProjectRequest.ValidateAll() if the designated
// constraints aren't met.
type DeployProjectRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeployProjectRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeployProjectRequestMultiError) AllErrors() []error { return m }

// DeployProjectRequestValidationError is the validation error returned by
// DeployProjectRequest.Validate if the designated constraints aren't met.
type DeployProjectRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeployProjectRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeployProjectRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeployProjectRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeployProjectRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeployProjectRequestValidationError) ErrorName() string {
	return "DeployProjectRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DeployProjectRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeployProjectRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeployProjectRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeployProjectRequestValidationError{}

// Validate checks the field values on DeployProjectResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DeployProjectResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DeployProjectResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DeployProjectResponseMultiError, or nil if none found.
func (m *DeployProjectResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DeployProjectResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for DeployId

	// no validation rules for Org

	// no validation rules for Project

	// no validation rules for FrontendUrl

	if len(errors) > 0 {
		return DeployProjectResponseMultiError(errors)
	}

	return nil
}

// DeployProjectResponseMultiError is an error wrapping multiple validation
// errors returned by DeployProjectResponse.ValidateAll() if the designated
// constraints aren't met.
type DeployProjectResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DeployProjectResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DeployProjectResponseMultiError) AllErrors() []error { return m }

// DeployProjectResponseValidationError is the validation error returned by
// DeployProjectResponse.Validate if the designated constraints aren't met.
type DeployProjectResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DeployProjectResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DeployProjectResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DeployProjectResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DeployProjectResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DeployProjectResponseValidationError) ErrorName() string {
	return "DeployProjectResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DeployProjectResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDeployProjectResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DeployProjectResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DeployProjectResponseValidationError{}

// Validate checks the field values on RedeployProjectRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *RedeployProjectRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RedeployProjectRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// RedeployProjectRequestMultiError, or nil if none found.
func (m *RedeployProjectRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *RedeployProjectRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for ProjectId

	// no validation rules for Reupload

	// no validation rules for Rearchive

	if len(errors) > 0 {
		return RedeployProjectRequestMultiError(errors)
	}

	return nil
}

// RedeployProjectRequestMultiError is an error wrapping multiple validation
// errors returned by RedeployProjectRequest.ValidateAll() if the designated
// constraints aren't met.
type RedeployProjectRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RedeployProjectRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RedeployProjectRequestMultiError) AllErrors() []error { return m }

// RedeployProjectRequestValidationError is the validation error returned by
// RedeployProjectRequest.Validate if the designated constraints aren't met.
type RedeployProjectRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RedeployProjectRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RedeployProjectRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RedeployProjectRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RedeployProjectRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RedeployProjectRequestValidationError) ErrorName() string {
	return "RedeployProjectRequestValidationError"
}

// Error satisfies the builtin error interface
func (e RedeployProjectRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRedeployProjectRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RedeployProjectRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RedeployProjectRequestValidationError{}

// Validate checks the field values on RedeployProjectResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *RedeployProjectResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RedeployProjectResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// RedeployProjectResponseMultiError, or nil if none found.
func (m *RedeployProjectResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *RedeployProjectResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for FrontendUrl

	if len(errors) > 0 {
		return RedeployProjectResponseMultiError(errors)
	}

	return nil
}

// RedeployProjectResponseMultiError is an error wrapping multiple validation
// errors returned by RedeployProjectResponse.ValidateAll() if the designated
// constraints aren't met.
type RedeployProjectResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RedeployProjectResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RedeployProjectResponseMultiError) AllErrors() []error { return m }

// RedeployProjectResponseValidationError is the validation error returned by
// RedeployProjectResponse.Validate if the designated constraints aren't met.
type RedeployProjectResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RedeployProjectResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RedeployProjectResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RedeployProjectResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RedeployProjectResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RedeployProjectResponseValidationError) ErrorName() string {
	return "RedeployProjectResponseValidationError"
}

// Error satisfies the builtin error interface
func (e RedeployProjectResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRedeployProjectResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RedeployProjectResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RedeployProjectResponseValidationError{}

// Validate checks the field values on GetCurrentUserRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetCurrentUserRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetCurrentUserRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetCurrentUserRequestMultiError, or nil if none found.
func (m *GetCurrentUserRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetCurrentUserRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GetCurrentUserRequestMultiError(errors)
	}

	return nil
}

// GetCurrentUserRequestMultiError is an error wrapping multiple validation
// errors returned by GetCurrentUserRequest.ValidateAll() if the designated
// constraints aren't met.
type GetCurrentUserRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetCurrentUserRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetCurrentUserRequestMultiError) AllErrors() []error { return m }

// GetCurrentUserRequestValidationError is the validation error returned by
// GetCurrentUserRequest.Validate if the designated constraints aren't met.
type GetCurrentUserRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetCurrentUserRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetCurrentUserRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetCurrentUserRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetCurrentUserRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetCurrentUserRequestValidationError) ErrorName() string {
	return "GetCurrentUserRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetCurrentUserRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetCurrentUserRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetCurrentUserRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetCurrentUserRequestValidationError{}

// Validate checks the field values on GetCurrentUserResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetCurrentUserResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetCurrentUserResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetCurrentUserResponseMultiError, or nil if none found.
func (m *GetCurrentUserResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetCurrentUserResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetUser()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetCurrentUserResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetCurrentUserResponseValidationError{
					field:  "User",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetUser()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetCurrentUserResponseValidationError{
				field:  "User",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for IsRepresentingUser

	if len(errors) > 0 {
		return GetCurrentUserResponseMultiError(errors)
	}

	return nil
}

// GetCurrentUserResponseMultiError is an error wrapping multiple validation
// errors returned by GetCurrentUserResponse.ValidateAll() if the designated
// constraints aren't met.
type GetCurrentUserResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetCurrentUserResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetCurrentUserResponseMultiError) AllErrors() []error { return m }

// GetCurrentUserResponseValidationError is the validation error returned by
// GetCurrentUserResponse.Validate if the designated constraints aren't met.
type GetCurrentUserResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetCurrentUserResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetCurrentUserResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetCurrentUserResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetCurrentUserResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetCurrentUserResponseValidationError) ErrorName() string {
	return "GetCurrentUserResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetCurrentUserResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetCurrentUserResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetCurrentUserResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetCurrentUserResponseValidationError{}

// Validate checks the field values on GetCurrentProjectRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetCurrentProjectRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetCurrentProjectRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetCurrentProjectRequestMultiError, or nil if none found.
func (m *GetCurrentProjectRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetCurrentProjectRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return GetCurrentProjectRequestMultiError(errors)
	}

	return nil
}

// GetCurrentProjectRequestMultiError is an error wrapping multiple validation
// errors returned by GetCurrentProjectRequest.ValidateAll() if the designated
// constraints aren't met.
type GetCurrentProjectRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetCurrentProjectRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetCurrentProjectRequestMultiError) AllErrors() []error { return m }

// GetCurrentProjectRequestValidationError is the validation error returned by
// GetCurrentProjectRequest.Validate if the designated constraints aren't met.
type GetCurrentProjectRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetCurrentProjectRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetCurrentProjectRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetCurrentProjectRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetCurrentProjectRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetCurrentProjectRequestValidationError) ErrorName() string {
	return "GetCurrentProjectRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetCurrentProjectRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetCurrentProjectRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetCurrentProjectRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetCurrentProjectRequestValidationError{}

// Validate checks the field values on GetCurrentProjectResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetCurrentProjectResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetCurrentProjectResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetCurrentProjectResponseMultiError, or nil if none found.
func (m *GetCurrentProjectResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetCurrentProjectResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for LocalProjectName

	if all {
		switch v := interface{}(m.GetProject()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetCurrentProjectResponseValidationError{
					field:  "Project",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetCurrentProjectResponseValidationError{
					field:  "Project",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetProject()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetCurrentProjectResponseValidationError{
				field:  "Project",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetCurrentProjectResponseMultiError(errors)
	}

	return nil
}

// GetCurrentProjectResponseMultiError is an error wrapping multiple validation
// errors returned by GetCurrentProjectResponse.ValidateAll() if the
// designated constraints aren't met.
type GetCurrentProjectResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetCurrentProjectResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetCurrentProjectResponseMultiError) AllErrors() []error { return m }

// GetCurrentProjectResponseValidationError is the validation error returned by
// GetCurrentProjectResponse.Validate if the designated constraints aren't met.
type GetCurrentProjectResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetCurrentProjectResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetCurrentProjectResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetCurrentProjectResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetCurrentProjectResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetCurrentProjectResponseValidationError) ErrorName() string {
	return "GetCurrentProjectResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetCurrentProjectResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetCurrentProjectResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetCurrentProjectResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetCurrentProjectResponseValidationError{}

// Validate checks the field values on
// ListOrganizationsAndBillingMetadataRequest with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ListOrganizationsAndBillingMetadataRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on
// ListOrganizationsAndBillingMetadataRequest with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in
// ListOrganizationsAndBillingMetadataRequestMultiError, or nil if none found.
func (m *ListOrganizationsAndBillingMetadataRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListOrganizationsAndBillingMetadataRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListOrganizationsAndBillingMetadataRequestMultiError(errors)
	}

	return nil
}

// ListOrganizationsAndBillingMetadataRequestMultiError is an error wrapping
// multiple validation errors returned by
// ListOrganizationsAndBillingMetadataRequest.ValidateAll() if the designated
// constraints aren't met.
type ListOrganizationsAndBillingMetadataRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListOrganizationsAndBillingMetadataRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListOrganizationsAndBillingMetadataRequestMultiError) AllErrors() []error { return m }

// ListOrganizationsAndBillingMetadataRequestValidationError is the validation
// error returned by ListOrganizationsAndBillingMetadataRequest.Validate if
// the designated constraints aren't met.
type ListOrganizationsAndBillingMetadataRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListOrganizationsAndBillingMetadataRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListOrganizationsAndBillingMetadataRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListOrganizationsAndBillingMetadataRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListOrganizationsAndBillingMetadataRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListOrganizationsAndBillingMetadataRequestValidationError) ErrorName() string {
	return "ListOrganizationsAndBillingMetadataRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListOrganizationsAndBillingMetadataRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListOrganizationsAndBillingMetadataRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListOrganizationsAndBillingMetadataRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListOrganizationsAndBillingMetadataRequestValidationError{}

// Validate checks the field values on
// ListOrganizationsAndBillingMetadataResponse with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *ListOrganizationsAndBillingMetadataResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on
// ListOrganizationsAndBillingMetadataResponse with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in
// ListOrganizationsAndBillingMetadataResponseMultiError, or nil if none found.
func (m *ListOrganizationsAndBillingMetadataResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListOrganizationsAndBillingMetadataResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetOrgs() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListOrganizationsAndBillingMetadataResponseValidationError{
						field:  fmt.Sprintf("Orgs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListOrganizationsAndBillingMetadataResponseValidationError{
						field:  fmt.Sprintf("Orgs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListOrganizationsAndBillingMetadataResponseValidationError{
					field:  fmt.Sprintf("Orgs[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListOrganizationsAndBillingMetadataResponseMultiError(errors)
	}

	return nil
}

// ListOrganizationsAndBillingMetadataResponseMultiError is an error wrapping
// multiple validation errors returned by
// ListOrganizationsAndBillingMetadataResponse.ValidateAll() if the designated
// constraints aren't met.
type ListOrganizationsAndBillingMetadataResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListOrganizationsAndBillingMetadataResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListOrganizationsAndBillingMetadataResponseMultiError) AllErrors() []error { return m }

// ListOrganizationsAndBillingMetadataResponseValidationError is the validation
// error returned by ListOrganizationsAndBillingMetadataResponse.Validate if
// the designated constraints aren't met.
type ListOrganizationsAndBillingMetadataResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListOrganizationsAndBillingMetadataResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListOrganizationsAndBillingMetadataResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListOrganizationsAndBillingMetadataResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListOrganizationsAndBillingMetadataResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListOrganizationsAndBillingMetadataResponseValidationError) ErrorName() string {
	return "ListOrganizationsAndBillingMetadataResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListOrganizationsAndBillingMetadataResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListOrganizationsAndBillingMetadataResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListOrganizationsAndBillingMetadataResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListOrganizationsAndBillingMetadataResponseValidationError{}

// Validate checks the field values on CreateOrganizationRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateOrganizationRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateOrganizationRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateOrganizationRequestMultiError, or nil if none found.
func (m *CreateOrganizationRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateOrganizationRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetName()) < 1 {
		err := CreateOrganizationRequestValidationError{
			field:  "Name",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Description

	// no validation rules for DisplayName

	if len(errors) > 0 {
		return CreateOrganizationRequestMultiError(errors)
	}

	return nil
}

// CreateOrganizationRequestMultiError is an error wrapping multiple validation
// errors returned by CreateOrganizationRequest.ValidateAll() if the
// designated constraints aren't met.
type CreateOrganizationRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateOrganizationRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateOrganizationRequestMultiError) AllErrors() []error { return m }

// CreateOrganizationRequestValidationError is the validation error returned by
// CreateOrganizationRequest.Validate if the designated constraints aren't met.
type CreateOrganizationRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateOrganizationRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateOrganizationRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateOrganizationRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateOrganizationRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateOrganizationRequestValidationError) ErrorName() string {
	return "CreateOrganizationRequestValidationError"
}

// Error satisfies the builtin error interface
func (e CreateOrganizationRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateOrganizationRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateOrganizationRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateOrganizationRequestValidationError{}

// Validate checks the field values on CreateOrganizationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *CreateOrganizationResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on CreateOrganizationResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// CreateOrganizationResponseMultiError, or nil if none found.
func (m *CreateOrganizationResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *CreateOrganizationResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetOrganization()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, CreateOrganizationResponseValidationError{
					field:  "Organization",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, CreateOrganizationResponseValidationError{
					field:  "Organization",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetOrganization()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return CreateOrganizationResponseValidationError{
				field:  "Organization",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return CreateOrganizationResponseMultiError(errors)
	}

	return nil
}

// CreateOrganizationResponseMultiError is an error wrapping multiple
// validation errors returned by CreateOrganizationResponse.ValidateAll() if
// the designated constraints aren't met.
type CreateOrganizationResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m CreateOrganizationResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m CreateOrganizationResponseMultiError) AllErrors() []error { return m }

// CreateOrganizationResponseValidationError is the validation error returned
// by CreateOrganizationResponse.Validate if the designated constraints aren't met.
type CreateOrganizationResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e CreateOrganizationResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e CreateOrganizationResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e CreateOrganizationResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e CreateOrganizationResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e CreateOrganizationResponseValidationError) ErrorName() string {
	return "CreateOrganizationResponseValidationError"
}

// Error satisfies the builtin error interface
func (e CreateOrganizationResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sCreateOrganizationResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = CreateOrganizationResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = CreateOrganizationResponseValidationError{}

// Validate checks the field values on ListMatchingProjectsRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListMatchingProjectsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListMatchingProjectsRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListMatchingProjectsRequestMultiError, or nil if none found.
func (m *ListMatchingProjectsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListMatchingProjectsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListMatchingProjectsRequestMultiError(errors)
	}

	return nil
}

// ListMatchingProjectsRequestMultiError is an error wrapping multiple
// validation errors returned by ListMatchingProjectsRequest.ValidateAll() if
// the designated constraints aren't met.
type ListMatchingProjectsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListMatchingProjectsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListMatchingProjectsRequestMultiError) AllErrors() []error { return m }

// ListMatchingProjectsRequestValidationError is the validation error returned
// by ListMatchingProjectsRequest.Validate if the designated constraints
// aren't met.
type ListMatchingProjectsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListMatchingProjectsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListMatchingProjectsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListMatchingProjectsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListMatchingProjectsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListMatchingProjectsRequestValidationError) ErrorName() string {
	return "ListMatchingProjectsRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListMatchingProjectsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListMatchingProjectsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListMatchingProjectsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListMatchingProjectsRequestValidationError{}

// Validate checks the field values on ListMatchingProjectsResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListMatchingProjectsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListMatchingProjectsResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListMatchingProjectsResponseMultiError, or nil if none found.
func (m *ListMatchingProjectsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListMatchingProjectsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetProjects() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListMatchingProjectsResponseValidationError{
						field:  fmt.Sprintf("Projects[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListMatchingProjectsResponseValidationError{
						field:  fmt.Sprintf("Projects[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListMatchingProjectsResponseValidationError{
					field:  fmt.Sprintf("Projects[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListMatchingProjectsResponseMultiError(errors)
	}

	return nil
}

// ListMatchingProjectsResponseMultiError is an error wrapping multiple
// validation errors returned by ListMatchingProjectsResponse.ValidateAll() if
// the designated constraints aren't met.
type ListMatchingProjectsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListMatchingProjectsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListMatchingProjectsResponseMultiError) AllErrors() []error { return m }

// ListMatchingProjectsResponseValidationError is the validation error returned
// by ListMatchingProjectsResponse.Validate if the designated constraints
// aren't met.
type ListMatchingProjectsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListMatchingProjectsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListMatchingProjectsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListMatchingProjectsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListMatchingProjectsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListMatchingProjectsResponseValidationError) ErrorName() string {
	return "ListMatchingProjectsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListMatchingProjectsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListMatchingProjectsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListMatchingProjectsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListMatchingProjectsResponseValidationError{}

// Validate checks the field values on ListProjectsForOrgRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListProjectsForOrgRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListProjectsForOrgRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListProjectsForOrgRequestMultiError, or nil if none found.
func (m *ListProjectsForOrgRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListProjectsForOrgRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if utf8.RuneCountInString(m.GetOrg()) < 1 {
		err := ListProjectsForOrgRequestValidationError{
			field:  "Org",
			reason: "value length must be at least 1 runes",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if m.GetPageSize() != 0 {

		if m.GetPageSize() > 1000 {
			err := ListProjectsForOrgRequestValidationError{
				field:  "PageSize",
				reason: "value must be less than or equal to 1000",
			}
			if !all {
				return err
			}
			errors = append(errors, err)
		}

	}

	// no validation rules for PageToken

	if len(errors) > 0 {
		return ListProjectsForOrgRequestMultiError(errors)
	}

	return nil
}

// ListProjectsForOrgRequestMultiError is an error wrapping multiple validation
// errors returned by ListProjectsForOrgRequest.ValidateAll() if the
// designated constraints aren't met.
type ListProjectsForOrgRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListProjectsForOrgRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListProjectsForOrgRequestMultiError) AllErrors() []error { return m }

// ListProjectsForOrgRequestValidationError is the validation error returned by
// ListProjectsForOrgRequest.Validate if the designated constraints aren't met.
type ListProjectsForOrgRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListProjectsForOrgRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListProjectsForOrgRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListProjectsForOrgRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListProjectsForOrgRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListProjectsForOrgRequestValidationError) ErrorName() string {
	return "ListProjectsForOrgRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListProjectsForOrgRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListProjectsForOrgRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListProjectsForOrgRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListProjectsForOrgRequestValidationError{}

// Validate checks the field values on ListProjectsForOrgResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListProjectsForOrgResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListProjectsForOrgResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListProjectsForOrgResponseMultiError, or nil if none found.
func (m *ListProjectsForOrgResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListProjectsForOrgResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetProjects() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListProjectsForOrgResponseValidationError{
						field:  fmt.Sprintf("Projects[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListProjectsForOrgResponseValidationError{
						field:  fmt.Sprintf("Projects[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListProjectsForOrgResponseValidationError{
					field:  fmt.Sprintf("Projects[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListProjectsForOrgResponseMultiError(errors)
	}

	return nil
}

// ListProjectsForOrgResponseMultiError is an error wrapping multiple
// validation errors returned by ListProjectsForOrgResponse.ValidateAll() if
// the designated constraints aren't met.
type ListProjectsForOrgResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListProjectsForOrgResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListProjectsForOrgResponseMultiError) AllErrors() []error { return m }

// ListProjectsForOrgResponseValidationError is the validation error returned
// by ListProjectsForOrgResponse.Validate if the designated constraints aren't met.
type ListProjectsForOrgResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListProjectsForOrgResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListProjectsForOrgResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListProjectsForOrgResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListProjectsForOrgResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListProjectsForOrgResponseValidationError) ErrorName() string {
	return "ListProjectsForOrgResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListProjectsForOrgResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListProjectsForOrgResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListProjectsForOrgResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListProjectsForOrgResponseValidationError{}

// Validate checks the field values on GetProjectRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetProjectRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetProjectRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetProjectRequestMultiError, or nil if none found.
func (m *GetProjectRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetProjectRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for OrganizationName

	// no validation rules for Name

	if len(errors) > 0 {
		return GetProjectRequestMultiError(errors)
	}

	return nil
}

// GetProjectRequestMultiError is an error wrapping multiple validation errors
// returned by GetProjectRequest.ValidateAll() if the designated constraints
// aren't met.
type GetProjectRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetProjectRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetProjectRequestMultiError) AllErrors() []error { return m }

// GetProjectRequestValidationError is the validation error returned by
// GetProjectRequest.Validate if the designated constraints aren't met.
type GetProjectRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetProjectRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetProjectRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetProjectRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetProjectRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetProjectRequestValidationError) ErrorName() string {
	return "GetProjectRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetProjectRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetProjectRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetProjectRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetProjectRequestValidationError{}

// Validate checks the field values on GetProjectResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetProjectResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetProjectResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetProjectResponseMultiError, or nil if none found.
func (m *GetProjectResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetProjectResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetProject()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetProjectResponseValidationError{
					field:  "Project",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetProjectResponseValidationError{
					field:  "Project",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetProject()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetProjectResponseValidationError{
				field:  "Project",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetProjectPermissions()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetProjectResponseValidationError{
					field:  "ProjectPermissions",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetProjectResponseValidationError{
					field:  "ProjectPermissions",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetProjectPermissions()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetProjectResponseValidationError{
				field:  "ProjectPermissions",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetProjectResponseMultiError(errors)
	}

	return nil
}

// GetProjectResponseMultiError is an error wrapping multiple validation errors
// returned by GetProjectResponse.ValidateAll() if the designated constraints
// aren't met.
type GetProjectResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetProjectResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetProjectResponseMultiError) AllErrors() []error { return m }

// GetProjectResponseValidationError is the validation error returned by
// GetProjectResponse.Validate if the designated constraints aren't met.
type GetProjectResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetProjectResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetProjectResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetProjectResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetProjectResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetProjectResponseValidationError) ErrorName() string {
	return "GetProjectResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetProjectResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetProjectResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetProjectResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetProjectResponseValidationError{}

// Validate checks the field values on
// ListOrganizationsAndBillingMetadataResponse_OrgMetadata with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListOrganizationsAndBillingMetadataResponse_OrgMetadata) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on
// ListOrganizationsAndBillingMetadataResponse_OrgMetadata with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListOrganizationsAndBillingMetadataResponse_OrgMetadataMultiError, or nil
// if none found.
func (m *ListOrganizationsAndBillingMetadataResponse_OrgMetadata) ValidateAll() error {
	return m.validate(true)
}

func (m *ListOrganizationsAndBillingMetadataResponse_OrgMetadata) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	for idx, item := range m.GetIssues() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError{
						field:  fmt.Sprintf("Issues[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError{
						field:  fmt.Sprintf("Issues[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError{
					field:  fmt.Sprintf("Issues[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListOrganizationsAndBillingMetadataResponse_OrgMetadataMultiError(errors)
	}

	return nil
}

// ListOrganizationsAndBillingMetadataResponse_OrgMetadataMultiError is an
// error wrapping multiple validation errors returned by
// ListOrganizationsAndBillingMetadataResponse_OrgMetadata.ValidateAll() if
// the designated constraints aren't met.
type ListOrganizationsAndBillingMetadataResponse_OrgMetadataMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListOrganizationsAndBillingMetadataResponse_OrgMetadataMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListOrganizationsAndBillingMetadataResponse_OrgMetadataMultiError) AllErrors() []error {
	return m
}

// ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError is
// the validation error returned by
// ListOrganizationsAndBillingMetadataResponse_OrgMetadata.Validate if the
// designated constraints aren't met.
type ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError) Field() string {
	return e.field
}

// Reason function returns reason value.
func (e ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError) Reason() string {
	return e.reason
}

// Cause function returns cause value.
func (e ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError) Cause() error {
	return e.cause
}

// Key function returns key value.
func (e ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError) Key() bool {
	return e.key
}

// ErrorName returns error name.
func (e ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError) ErrorName() string {
	return "ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError"
}

// Error satisfies the builtin error interface
func (e ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListOrganizationsAndBillingMetadataResponse_OrgMetadata.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListOrganizationsAndBillingMetadataResponse_OrgMetadataValidationError{}
